src/types.tsx

```diff
@@ -7,11 +7,29 @@ export type Field<T> = {
   touched: boolean
   error?: string
   rules: ValidationRule<T>[]
-  setValue: (value: T) => void
+  setValue: <
+    U extends SetValueOptions | EmptySetValueOptions = EmptySetValueOptions
+  >(
+    value: T,
+    options?: U
+  ) => SetValueReturnType<U>
   reset: () => void
   validate: () => void
 }
 
+/** The return type of the setValue function, which can vary based on the options passed in */
+export type SetValueReturnType<T> = T extends EmptySetValueOptions
+  ? void
+  : Promise<boolean>
+
+export type EmptySetValueOptions = {
+  __type: 'empty'
```

Comment:
So WTF is this?

Well recall that TypeScript is _structurally_ typed, not _nominally_ typed -- i.e. type equality is calculated not on name (like most languages), but on _shape_ (same field names / types). 

What this `__type: 'empty'` does is basically give us a "phantom tag" that allows us to tell TS  this type is totally unique. And we want to do that so TS doesn't think our `EmptySetValueOptions` extends `SetValueOptions` (or vice versa). Doing this ensures our `extends` clause on line 21 works as expected. 

I tend to refer to these as "phantom tags" as they are only used at compile time convince TS that one type distinct from another.